615 Creating a Tutorial for the CEP-78 JS Client (#92)
casper-ecosystem/enhanced-nft: client-js/TUTORIAL.md

# CEP-78 JavaScript Client Tutorial

This tutorial outlines the usage of the JavaScript client available for the CEP-78 Enhanced NFT Standard.

> Further information on the CEP-78 Enhanced NFT Standard can be found [here](https://github.com/casper-ecosystem/cep-78-enhanced-nft).

Reviewer: ```suggestion
Further information on the CEP-78 Enhanced NFT Standard can be found [here](https://github.com/casper-ecosystem/cep-78-enhanced-nft).

The client is available in the npm as [casper-cep78-js-client](https://www.npmjs.com/package/casper-cep78-js-client).
```

Reviewer: The link is not working rn, but it will after we publish it :)

> ## Installing Dependencies
> 
> Dependencies can be installed using the following command in the `client-js` directory:
> 
> ```js
> npm i
> ```

Reviewer: let's replace the whole section with 
```suggestion
## Client Instalation

The client can be installed in your project that is built using TypeScript / JavaScript.
To install run:

```js
npm install casper-cep78-js-client
```
```

      metadataMutability: MetadataMutability.Immutable,
      mintingMode: MintingMode.Installer,
    },
    "165000000000",
    FAUCET_KEYS.publicKey,
    [FAUCET_KEYS]
  );

  const hash = await installDeploy.send(process.env.NODE_URL!);

```

`InstallArgs` are specified sd follows:

* `collectionName` - The name of the NFT collection, passed in as a `String`. **This parameter is required and cannot be changed post installation**.

* `collectionSymbol` - The symbol representing a given NFT collection, passed in as a `String`. **This parameter is required and cannot be changed post installation**.

* `totalTokenSupply` - The total number of NFTs that a specific contract instance will mint passed in as a `U64` value. **This parameter is required and cannot be changed post installation**.

* `ownershipMode` - The `OwnershipMode` modality that dictates the ownership behavior of the NFT contract. This argument is passed in as a `u8` value and is required at the time of installation.

* `nftKind` - The `NFTKind` modality that specifies the off-chain items represented by the on-chain NFT data. This argument is passed in as a `u8` value and is required at the time of installation.

> * `jsonSchema` - The JSON schema for the NFT tokens that will be minted by the NFT contract passed in as a `String`. This parameter is required if the metadata kind is set to `CustomValidated(4)` and **cannot be changed post installation**.

Reviewer 2: Hi, I believe this parameter is also required for 0/1/2/3 metadata kind but it can be empty for Raw(3) only. CustomValidated(4) requires however a specific formatted schema.

    [FAUCET_KEYS]
  );

  const mintDeployHash = await mintDeploy.send(NODE_URL!);

```
The arguments adhere to those provided in the original installation.

## Transferring a Token

After minting one or more tokens, you can then use the following code to transfer the tokens between accounts:

```js

  const transferDeploy = await contractClient.transfer(
    {
      tokenId: "0",
      source: FAUCET_KEYS.publicKey,
      target: USER1_KEYS.publicKey,
    },
    "13000000000",
    FAUCET_KEYS.publicKey,
    [FAUCET_KEYS]
  );

  const transferDeployHash = await transferDeploy.send(NODE_URL!);

```

Transferring accepts the following arguments:

* `tokenId` - The sequential ID assigned to a token in mint order.

* `source` - The account sending the token in question.

* `target` - The account receiving the transferred token.

## Burning a Token

The following code shows how to burn a minted NFT that you hold and have access rights to, requiring only the `tokenId` argument:

```js

  const burnDeploy = await contractClient.burn(
    { tokenId: "0" },
    "13000000000",
    USER1_KEYS.publicKey,
    [USER1_KEYS]
  );

  const burnDeployHash = await burnDeploy.send(NODE_URL!);

```

> ## Testing

Reviewer: ```suggestion
## Example usages
```

> ### Running an Install Test

Reviewer: ```suggestion
### Running an Install Example
```

This repository includes a test script for installing a CEP-78 contract instance.

You will need to define the following variables in the `.env` file:

* `NODE_URL` - The address of a node. If you are testing using [NCTL](https://docs.casperlabs.io/dapp-dev-guide/building-dapps/setup-nctl/), this will be `http://localhost:11101/rpc`.

* `NETWORK_NAME` - The name of the Casper network you are testing on, `casper-net-1` when testing using a local network with NCTL.

* `MASTER_KEY_PAIR_PATH` - The path to the key pair of the minting account.

* `USER1_KEY_PAIR_PATH` - The path to an additional account's key pair for use in testing transfer features.

You may also need to install associated dependencies using:

```js
npm i
```

This test can be run using the following command:

```js

> npm run test:install

Reviewer: ```suggestion
npm run example:install
```

```

The test will then return the installation's `deployHash`, and inform you when the installation is successful.

The test will then provide the installing account's information, which will include the CEP-78 NFT contract's hash and package hash.

### Running a Usage Test

A usage test uses the same variables as the Install test above, but tests the basic functionality of the contract after installation.

The usage test can be run using the following command:

```js

> npm run test:usage

Reviewer: ```suggestion
npm run example:usage
```

```

> This test will acquire the contract's hash and package hash, prior to sending three separate deploys to perform several function tests as follows:

Reviewer: In this section I would replace `test` for `example` everywhere. Also we can point users that they are running code that is in `client-js/examples` directory so they can treat it as reference :)
